src/Authentication/Authentication/Helpers/EncodeUriPath.cs

```diff
@@ -0,0 +1,34 @@
+// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+using System;
+using System.Text;
+
+namespace Microsoft.Graph.PowerShell.Authentication.Helpers
+{
+    /// <summary>
+    /// Escape data string/url encode Uris that have paths containing special characters like #.
+    /// For a path like /beta/users/first.last_foo.com#EXT#@contoso.onmicrosoft.com, the last segment contains special characters that need to be escaped
+    /// </summary>
+    public static class EncodeUriPaths
+    {
+        public static Uri EscapeDataStrings(this Uri uri)
+        {
+            int counter = 0;
+            var pathSegments = uri.OriginalString.Split('/');
+            StringBuilder sb = new StringBuilder();
+            foreach (var segment in pathSegments)
+            {
```

Comment:
Why not use `uri.Segments` instead of `uri.OriginalString.Split('/')`?

```suggestion
            StringBuilder sb = new StringBuilder();
            foreach (var segment in uri.Segments)
            {

```

Reply:
Using uri.Segments will ommit the part with special characters.  Given a url like this https://graph.microsoft.com/beta/users/first.last_foo.com#EXT#@contoso.onmicrosoft.com., You'll only end up getting '/beta/users/first.last_foo.com' yet we want the whole thing including that omitted part with special characters.

Reply:
It makes sense. So how do you avoid swallowing anchor components in a URL? Because `#` marks the beginning of an anchor.

Reply:
good call @calebkiage, turns out this was a problem.